### PR TITLE
bugfix(EmotionFX): update Motion Event Inspector when event data has changed

### PR DESCRIPTION
**Pull request by pollend**
ref: https://github.com/o3de/o3de/issues/12665
Signed-off-by: Michael Pollind <mpollind@gmail.com>

## What does this PR do?

I added some logic to update the event data when its changed. not sure if this is the best way to approach this, but I think this should handle the case where an event is update remotely/deleted since this logic will just re-query the data from TimerViewerPlugin.cpp.

There is quite a lot here, I think can be refactored. This can be pretty hard to follow. 

## How was this PR tested?

follows steps in linked PR.

### CODE REVIEW

File: Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/MotionEvents/MotionEventWidget.cpp [hunk 1]
```diff
@@ -36,9 +40,13 @@ namespace EMStudio
...
 
-
     void MotionEventWidget::ReInit(EMotionFX::Motion* motion, EMotionFX::MotionEvent* motionEvent)
     {
+        m_changeEventHandler.Disconnect();
+        if (motionEvent)
+        {
+            motionEvent->SetEventDataChangeEvent(m_changeEventHandler);
+        }
         m_editor->SetMotionEvent(motion, motionEvent);
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/MotionEvents/MotionEventWidget.cpp:50]: question: Should this be moved inside the null check too?

File: Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/TimeView/TimeViewPlugin.h [hunk 2]
```diff
@@ -182,6 +182,7 @@ namespace EMStudio
         void MotionEventTrackChanged(size_t eventNr, float startTime, float endTime, const char* oldTrackName, const char* newTrackName)            { UnselectAllElements(); CommandSystem::CommandHelperMotionEventTrackChanged(eventNr, startTime, endTime, oldTrackName, newTrackName); }
         void OnManualTimeChange(float timeValue);
         void toggleMotionEventPresetsPane();
+        void onRefreshSelection();
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/TimeView/TimeViewPlugin.h:185]: update: I think as we're not overriding here this should be of the form `OnRefreshSelection` to be consistent with the coding standards

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.h [hunk 3]
```diff
@@ -42,7 +45,21 @@ namespace EMotionFX
...
+        void SetEventDataChangeEvent(Event::EventDataChangeEvent::Handler& handler);
     protected:
+        struct EventContainer {
+            EventContainer() = default;
+             
+            // event container is not copyable 
+            EventContainer(const EventContainer&) {}
+            EventContainer& operator=(const EventContainer&) {
+                return *this;
+            }
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Source/Event.h:57]: update: This doesn't look correct to me, these operations are not correct, should we `=delete` them if they can't be used. Currently they'll run but do the wrong thing (not actually copy anything)

**Review by hultonha — COMMENTED** [review]
Thanks very much for taking a look at this @pollend, there are a couple of things I've got questions about it would be good to sanity check. Would you also please be able to include a video of how the operation behaves now to show it compared to that described in the bug. Thanks!

**Comment by amzn-rhhong** [comment]
I will review this later this week. Awesome work @pollend

File: Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/MotionEvents/MotionEventWidget.cpp (hunk 1 above)
Comment (pollend) [reply Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/MotionEvents/MotionEventWidget.cpp:50]: This is the existing logic. It will clear of the data  is null. The api is not clear so we will have to revisit this I think

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.h (hunk 3 above)
Comment (pollend) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h:57]: I'm cheating a bit event is not a copyable type is in stubbing the copy and assignment.

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.h [hunk 4]
```diff
@@ -42,7 +45,21 @@ namespace EMotionFX
         void SetEventData(size_t index, EventDataPtr&& newData);
         void InsertEventData(size_t index, EventDataPtr&& newData);
 
+        void SetEventDataChangeEvent(Event::EventDataChangeEvent::Handler& handler);
     protected:
+        struct EventContainer {
```
[the remark this answers is not in the document]
Comment (amzn-rhhong) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h:50]: Why are you copying an event? I think AZ::event aren't supposed to be copied by design. EventHanderA is suppose to only handle EventA. If there is an EventB that copied from EventA, it will trigger undefined behavior.

[the remark this answers is not in the document]
Comment (pollend) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h:50]: I'm not. The events are discarded when the emfx event is copied. It's done in a couple places.

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.h (hunk 3 above)
Comment (hultonha) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h:57]: Yeah I don't think we should do this, there definitely has to be another option

Comment (hultonha) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h:57]: I think in this case we should provide a proper implementation of the Event copy/move constructor and move/assignment operator to correctly clone the event and its callback, rather than this cheat. This is much like the same thing as `std::unique_ptr` which isn't copyable but you can of course do a deep copy of the underlying value if supported which I think should be possible here right? Or we need to make `Event` itself non-copyable

Comment (pollend) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h:57]: I end up having the implement the copy constructor for both Event and MotionEvent I think just cheating a bit here but yes this is probably the correct thing I should of done originally.

Comment (hultonha) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h]: Yeah in cases like this the rule of 5 is the way to go (implement all special members) - you can use the `swap` idiom to implement some in terms of the others but it's up to you what you decide - this talk on the special members is really great too if you haven't seen it - https://www.youtube.com/watch?v=vLinb2fgkHk.

Comment (pollend) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h]: yep, I ended up implementing all 5 operators.

Comment (hultonha) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.h]: Okay excellent thanks @pollend, I'll take another look shortly

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.cpp [hunk 5]
```diff
@@ -25,6 +36,18 @@ namespace EMotionFX
...
+    {
+        m_eventDatas = other.m_eventDatas;
+        return *this;
+    }
+
+    Event& Event::operator=(Event&& other) 
+    {
+        m_eventDatas = AZStd::move(other.m_eventDatas);
+        return *this;
+    }
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Source/Event.cpp:50]: It looks like `m_eventDatasChangeEvent` is missing here?

To be honest we really should write some tests for this logic to ensure all the right things are getting copied/moved as expected otherwise we won't be sure this is working correctly

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.h [hunk 6]
```diff
@@ -42,7 +48,15 @@ namespace EMotionFX
         void SetEventData(size_t index, EventDataPtr&& newData);
         void InsertEventData(size_t index, EventDataPtr&& newData);
 
+        void SetEventDataChangeEvent(Event::EventDataChangeEvent::Handler& handler);
+
+        Event& operator=(const Event& other);
+
+        Event& operator=(Event&& other);
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Source/Event.h:55]: update: Please can you group these with the other special members

File: Gems/EMotionFX/Code/EMotionFX/Source/MotionEvent.cpp [hunk 7]
```diff
@@ -30,6 +30,23 @@ namespace EMotionFX
...
+    MotionEvent::MotionEvent(const MotionEvent& event):
+        Event(event),
+        m_startTime(event.m_startTime),
+        m_endTime(event.m_endTime),
+        m_isSyncEvent(event.m_isSyncEvent)
+    {
+
+    }
+    MotionEvent::MotionEvent(MotionEvent&& event):
+        Event(event),
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Source/MotionEvent.cpp:42]: update: You need to add `AZStd::move` to this parameter (`event`) otherwise it won't be moved

File: Gems/EMotionFX/Code/EMotionFX/Source/MotionEvent.cpp [hunk 8]
```diff
@@ -62,6 +79,24 @@ namespace EMotionFX
...
+        Event::operator=(other);
+        m_startTime = other.m_startTime;
+        m_endTime = other.m_endTime;
+        m_isSyncEvent = other.m_isSyncEvent;
+        return *this;
+    }
+
+    MotionEvent& MotionEvent::operator=(MotionEvent&& other)
+    {
+        Event::operator=(other);
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Source/MotionEvent.cpp:93]: update: Need to `AZStd::move` `other` here too otherwise it will call the normal assignment operator - as this stuff is complex we really should add tests to cover it

File: Gems/EMotionFX/Code/EMotionFX/Source/MotionEvent.h [hunk 9]
```diff
@@ -96,6 +98,10 @@ namespace EMotionFX
          */
         MotionEvent(float startTimeValue, float endTimeValue, EventDataSet&& datas);
 
+        MotionEvent& operator=(const MotionEvent& other);
+        
+        MotionEvent& operator=(MotionEvent&& other);
```
Comment (hultonha) [inline Gems/EMotionFX/Code/EMotionFX/Source/MotionEvent.h:103]: update: Please group these with the other special member calls

File: Gems/EMotionFX/Code/EMotionFX/Source/Event.cpp [hunk 10]
```diff
@@ -16,6 +16,17 @@ namespace EMotionFX
 {
     AZ_CLASS_ALLOCATOR_IMPL(Event, MotionEventAllocator, 0)
 
+    Event::Event(const Event& data)
+        : m_eventDatas(data.m_eventDatas)
+    {
+    }
```
[the remark this answers is not in the document]
Comment (hultonha) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.cpp:22]: That would be really great, thank you!

[the remark this answers is not in the document]
Comment (pollend) [reply Gems/EMotionFX/Code/EMotionFX/Source/Event.cpp:22]: you can't clone `m_eventDatasChangeEvent`

**Comment by hultonha** [comment]
After taking a closer look at this issue @pollend I'm closing in favour of https://github.com/o3de/o3de/pull/13383 which keeps all changes at a slightly higher level and does not require changes to `Motion` or `MotionEvent`. I think this is a simpler approach and achieves the same ends. Let me know what you think. Thanks!
